File: sdks/sandbox/python/src/opensandbox/mount/_shell.py
#
# Copyright 2025 Alibaba Group Holding Ltd.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
"""
Internal helpers that translate :mod:`opensandbox.mount.models` requests into
shell commands and validate execution results.

These helpers are shared by the async and sync mount implementations so both
SDK entry points build identical commands.
"""

from __future__ import annotations

import shlex
import uuid
from dataclasses import dataclass

from opensandbox.exceptions import InvalidArgumentException, MountFailedException
from opensandbox.models.execd import Execution
from opensandbox.models.filesystem import WriteEntry
from opensandbox.mount.models import (
    DEFAULT_NFS_OPTIONS,
    NfsMountOptions,
    OssfsMountOptions,
    OssfsVersion,
)

_OSSFS1_PASSWD_PREFIX = "/tmp/opensandbox-ossfspass-"
"""Prefix for the per-call ossfs 1.x password file path under /tmp."""


@dataclass(frozen=True)
class Ossfs2Plan:
    """
    Plan for an ossfs 2.x mount split into a configuration file that must be
    uploaded first and the actual mount command to run.
    """

    conf_path: str
    conf_content: str
    command: str


def validate_nfs(options: NfsMountOptions) -> None:
    """Validate NFS options before assembling the mount command."""
    if not options.endpoint or not options.endpoint.strip():
        raise InvalidArgumentException("endpoint must not be blank")
    if not options.mount_point or not options.mount_point.strip():
        raise InvalidArgumentException("mount_point must not be blank")
    if not options.nas_path or not options.nas_path.strip():
        raise InvalidArgumentException("nas_path must not be blank")


def validate_ossfs(options: OssfsMountOptions) -> None:
    """Validate OSS options before assembling the mount command."""
    if not options.endpoint or not options.endpoint.strip():
        raise InvalidArgumentException("endpoint must not be blank")
    if not options.bucket or not options.bucket.strip():
        raise InvalidArgumentException("bucket must not be blank")
    if not options.mount_point or not options.mount_point.strip():
        raise InvalidArgumentException("mount_point must not be blank")
    if not options.access_key_id or not options.access_key_id.strip():
        raise InvalidArgumentException("access_key_id must not be blank")
    if not options.access_key_secret or not options.access_key_secret.strip():
        raise InvalidArgumentException("access_key_secret must not be blank")


def build_nfs_command(options: NfsMountOptions) -> str:
    """
    Assemble the shell command that mounts an NFS export.

    Layout:
        [installation && ] mkdir -p <mnt> && mount -t nfs -o <opts> <server>:<path> <mnt>
    """
    opt_string = (
        options.options if options.options and options.options.strip() else DEFAULT_NFS_OPTIONS
    )
    source = f"{options.endpoint}:{options.nas_path}"
    core = (
        f"mkdir -p {shlex.quote(options.mount_point)} && "
        f"mount -t nfs -o {shlex.quote(opt_string)} "
        f"{shlex.quote(source)} {shlex.quote(options.mount_point)}"
    )
    return _prepend_installation(options.installation, core)


def build_ossfs1_command(options: OssfsMountOptions) -> str:
    """
    Assemble the shell command that mounts an OSS bucket with ossfs 1.x.

    Uses a unique per-call password file under ``/tmp`` so that concurrent
    ossfs 1.x mounts in the same sandbox do not overwrite or delete each
    other's credentials. The file is created atomically with mode 600
    (avoids the naive ``echo > file; chmod 600`` world-readable window) and
    is always removed afterwards regardless of whether ``ossfs`` itself
    succeeded.
    """
    passwd = f"{options.bucket}:{options.access_key_id}:{options.access_key_secret}"
    if options.security_token and options.security_token.strip():
        # STS mode: bucket:accessKeyId:accessKeySecret:securityToken
        passwd = f"{passwd}:{options.security_token}"

    bucket_arg = options.bucket
    if options.bucket_directory and options.bucket_directory.strip():
        bucket_arg = f"{options.bucket}:/{options.bucket_directory}"

    option_flags = "".join(f" -o{shlex.quote(opt)}" for opt in options.options)

    passwd_path = f"{_OSSFS1_PASSWD_PREFIX}{uuid.uuid4()}"
    quoted_passwd_path = shlex.quote(passwd_path)
    core = (
        "ossfs --version && "
        f"install -m 600 /dev/null {quoted_passwd_path} && "
        f"printf %s {shlex.quote(passwd)} > {quoted_passwd_path} && "
        f"mkdir -p {shlex.quote(options.mount_point)} && "
        f"( ossfs {shlex.quote(bucket_arg)} {shlex.quote(options.mount_point)} "
        f"-ourl={shlex.quote(options.endpoint)} "
        f"-opasswd_file={quoted_passwd_path}{option_flags} ); "
        f"__rc=$?; rm -f {quoted_passwd_path}; exit $__rc"
    )
    return _prepend_installation(options.installation, core)


def build_ossfs2_plan(options: OssfsMountOptions) -> Ossfs2Plan:
    """Build the configuration file content and mount command for ossfs 2.x."""
    lines = [
        f"--oss_endpoint={options.endpoint}",
        f"--oss_bucket={options.bucket}",
    ]
    if options.bucket_directory and options.bucket_directory.strip():
        prefix = options.bucket_directory.rstrip("/") + "/"
        lines.append(f"--oss_bucket_prefix={prefix}")
    for opt in options.options:
        lines.append(f"--{opt}")
    conf_content = "\n".join(lines) + "\n"

    conf_path = f"/tmp/opensandbox-ossfs-{uuid.uuid4()}.conf"

    sts_export = ""
    if options.security_token and options.security_token.strip():
        sts_export = f" && export OSS_SESSION_TOKEN={shlex.quote(options.security_token)}"

    core = (
        "ossfs2 --version && "
        f"mkdir -p {shlex.quote(options.mount_point)} && "
        f"export OSS_ACCESS_KEY_ID={shlex.quote(options.access_key_id)} && "
        f"export OSS_ACCESS_KEY_SECRET={shlex.quote(options.access_key_secret)}"
        f"{sts_export}"
        f" && ossfs2 mount {shlex.quote(options.mount_point)} -c {shlex.quote(conf_path)}"
    )
    command = _prepend_installation(options.installation, core)
    return Ossfs2Plan(conf_path=conf_path, conf_content=conf_content, command=command)


def build_ossfs2_conf_entry(plan: Ossfs2Plan) -> WriteEntry:
    """Convert an :class:`Ossfs2Plan` into a :class:`WriteEntry` for upload."""
    return WriteEntry(path=plan.conf_path, data=plan.conf_content, mode=600)


def build_umount_command(mount_point: str) -> str:
    """Assemble the shell command that unmounts a path."""
    if not mount_point or not mount_point.strip():
        raise InvalidArgumentException("mount_point must not be blank")
    return f"umount {shlex.quote(mount_point)}"


def select_ossfs_version(options: OssfsMountOptions) -> OssfsVersion:
    """Return the effective ossfs version, defaulting to ossfs 1.0."""
    return options.version or OssfsVersion.OSSFS_1_0


def ensure_success(execution: Execution, failure_prefix: str) -> None:
    """Raise :class:`MountFailedException` if the execution reported an error."""
    error = execution.error
    exit_code = execution.exit_code
    failed = error is not None or (exit_code is not None and exit_code != 0)
    if not failed:
        return

    stderr_text = "\n".join(msg.text for msg in execution.logs.stderr)
    parts: list[str] = []
    if error is not None:
        parts.append(f"[{error.name}] {error.value}")
    if stderr_text:
        parts.append(f"stderr={stderr_text}")
    detail = " | ".join(parts)
    message = f"{failure_prefix}: {detail}" if detail else failure_prefix
    raise MountFailedException(message=message, execution=execution)


def _prepend_installation(installation: str | None, core: str) -> str:
    if installation and installation.strip():
        return f"{installation} && {core}"
    return core

File: sdks/sandbox/javascript/src/mount/shell.ts
// Copyright 2026 Alibaba Group Holding Ltd.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import {
  InvalidArgumentException,
  MountFailedException,
} from "../core/exceptions.js";
import type { Execution } from "../models/execution.js";
import type { WriteEntry } from "../models/filesystem.js";
import { DEFAULT_NFS_OPTIONS, type NfsMountOptions, type OssfsMountOptions, type OssfsVersion } from "./models.js";

/**
 * Prefix used for the per-call ossfs 1.x password file. A UUID is appended per
 * mount so that concurrent ossfs 1.x mounts in the same sandbox do not
 * overwrite or delete each other's credentials.
 */
export const OSSFS1_PASSWD_PATH_PREFIX = "/tmp/opensandbox-ossfspass-";

/** Plan for an ossfs 2.x mount: a conf file to upload plus a mount command. */
export interface Ossfs2Plan {
  confPath: string;
  confContent: string;
  command: string;
}

/**
 * Quote a value for POSIX shell single-quoted context.
 *
 * `'` inside the value is escaped as `'\''`. The result is always safe to embed
 * as one argument to a `sh -c` string. Unlike Python's `shlex.quote`, this
 * always adds quotes so that behavior is consistent across engines.
 */
export function shQuote(value: string): string {
  return `'${value.replace(/'/g, "'\\''")}'`;
}

export function validateNfs(options: NfsMountOptions): void {
  if (!options.endpoint?.trim()) {
    throw new InvalidArgumentException({ message: "endpoint must not be blank" });
  }
  if (!options.mountPoint?.trim()) {
    throw new InvalidArgumentException({ message: "mountPoint must not be blank" });
  }
  if (!options.nasPath?.trim()) {
    throw new InvalidArgumentException({ message: "nasPath must not be blank" });
  }
}

export function validateOssfs(options: OssfsMountOptions): void {
  if (!options.endpoint?.trim()) {
    throw new InvalidArgumentException({ message: "endpoint must not be blank" });
  }
  if (!options.bucket?.trim()) {
    throw new InvalidArgumentException({ message: "bucket must not be blank" });
  }
  if (!options.mountPoint?.trim()) {
    throw new InvalidArgumentException({ message: "mountPoint must not be blank" });
  }
  if (!options.accessKeyId?.trim()) {
    throw new InvalidArgumentException({ message: "accessKeyId must not be blank" });
  }
  if (!options.accessKeySecret?.trim()) {
    throw new InvalidArgumentException({ message: "accessKeySecret must not be blank" });
  }
}

export function buildNfsCommand(options: NfsMountOptions): string {
  const optString = options.options?.trim() ? options.options : DEFAULT_NFS_OPTIONS;
  const source = `${options.endpoint}:${options.nasPath}`;
  const core =
    `mkdir -p ${shQuote(options.mountPoint)} && ` +
    `mount -t nfs -o ${shQuote(optString)} ` +
    `${shQuote(source)} ${shQuote(options.mountPoint)}`;
  return prependInstallation(options.installation, core);
}

export function buildOssfs1Command(options: OssfsMountOptions): string {
  let passwd = `${options.bucket}:${options.accessKeyId}:${options.accessKeySecret}`;
  if (options.securityToken?.trim()) {
    passwd = `${passwd}:${options.securityToken}`;
  }

  let bucketArg = options.bucket;
  if (options.bucketDirectory?.trim()) {
    bucketArg = `${options.bucket}:/${options.bucketDirectory}`;
  }

  const optionFlags = (options.options ?? [])
    .map((opt) => ` -o${shQuote(opt)}`)
    .join("");

  // Use a unique per-call password file under /tmp so that concurrent
  // ossfs 1.x mounts do not overwrite or delete each other's credentials.
  // Create it with mode 0600 in one atomic step (avoids the world-readable
  // window that a naive `echo > ; chmod 600` would create). Always clean it
  // up even if ossfs fails, by preserving the subshell exit code.
  const passwdPath = `${OSSFS1_PASSWD_PATH_PREFIX}${randomUuid()}`;
  const quotedPasswdPath = shQuote(passwdPath);
  const core =
    "ossfs --version && " +
    `install -m 600 /dev/null ${quotedPasswdPath} && ` +
    `printf %s ${shQuote(passwd)} > ${quotedPasswdPath} && ` +
    `mkdir -p ${shQuote(options.mountPoint)} && ` +
    `( ossfs ${shQuote(bucketArg)} ${shQuote(options.mountPoint)} ` +
    `-ourl=${shQuote(options.endpoint)} ` +
    `-opasswd_file=${quotedPasswdPath}${optionFlags} ); ` +
    `__rc=$?; rm -f ${quotedPasswdPath}; exit $__rc`;

  return prependInstallation(options.installation, core);
}

export function buildOssfs2Plan(options: OssfsMountOptions): Ossfs2Plan {
  const lines: string[] = [
    `--oss_endpoint=${options.endpoint}`,
    `--oss_bucket=${options.bucket}`,
  ];
  if (options.bucketDirectory?.trim()) {
    const prefix = options.bucketDirectory.replace(/\/+$/, "") + "/";
    lines.push(`--oss_bucket_prefix=${prefix}`);
  }
  for (const opt of options.options ?? []) {
    lines.push(`--${opt}`);
  }
  const confContent = lines.join("\n") + "\n";

  const confPath = `/tmp/opensandbox-ossfs-${randomUuid()}.conf`;

  const stsExport =
    options.securityToken?.trim()
      ? ` && export OSS_SESSION_TOKEN=${shQuote(options.securityToken)}`
      : "";

  const core =
    "ossfs2 --version && " +
    `mkdir -p ${shQuote(options.mountPoint)} && ` +
    `export OSS_ACCESS_KEY_ID=${shQuote(options.accessKeyId)} && ` +
    `export OSS_ACCESS_KEY_SECRET=${shQuote(options.accessKeySecret)}` +
    stsExport +
    ` && ossfs2 mount ${shQuote(options.mountPoint)} -c ${shQuote(confPath)}`;

  const command = prependInstallation(options.installation, core);
  return { confPath, confContent, command };
}

export function buildOssfs2ConfEntry(plan: Ossfs2Plan): WriteEntry {
  return {
    path: plan.confPath,
    data: plan.confContent,
    // Mode is serialized as a JSON number and parsed by execd as an octal
    // string (see components/execd/pkg/web/controller/utils.go); pass the
    // decimal literal 600 rather than the JavaScript octal literal 0o600.
    mode: 600,
  };
}

export function buildUmountCommand(mountPoint: string): string {
  if (!mountPoint?.trim()) {
    throw new InvalidArgumentException({ message: "mountPoint must not be blank" });
  }
  return `umount ${shQuote(mountPoint)}`;
}

export function selectOssfsVersion(options: OssfsMountOptions): OssfsVersion {
  const v = options.version;
  if (v == null) {
    return "1.0";
  }
  if (v !== "1.0" && v !== "2.0") {
    // TypeScript's structural typing means a plain JS caller (or any caller
    // that has widened the value to a string) can pass "1", "3.0", etc. Fail
    // fast rather than silently dispatching to ossfs2, matching the stricter
    // Kotlin / Python / Go / C# implementations.
    throw new InvalidArgumentException({
      message: `Unsupported ossfs version: ${String(v)} (expected "1.0" or "2.0")`,
    });
  }
  return v;
}

export function ensureSuccess(execution: Execution, failurePrefix: string): void {
  const error = execution.error;
  const exitCode = execution.exitCode;
  const failed = error != null || (exitCode != null && exitCode !== 0);
  if (!failed) {
    return;
  }
  const stderrText = execution.logs.stderr.map((m) => m.text).join("\n");
  const parts: string[] = [];
  if (error) {
    parts.push(`[${error.name}] ${error.value}`);
  }
  if (stderrText) {
    parts.push(`stderr=${stderrText}`);
  }
  const detail = parts.join(" | ");
  const message = detail ? `${failurePrefix}: ${detail}` : failurePrefix;
  throw new MountFailedException({ message, execution });
}

function prependInstallation(installation: string | undefined, core: string): string {
  if (installation?.trim()) {
    return `${installation} && ${core}`;
  }
  return core;
}

function randomUuid(): string {
  // Prefer Web Crypto (Node 20+ / browsers). Fall back to a compact hex string.
  const g = globalThis as { crypto?: { randomUUID?: () => string } };
  if (g.crypto && typeof g.crypto.randomUUID === "function") {
    return g.crypto.randomUUID();
  }
  const rand = Math.random().toString(16).slice(2) + Math.random().toString(16).slice(2);
  return rand.slice(0, 32);
}
